src/storage/src/source/metrics.rs

```diff
@@ -196,6 +204,130 @@ impl PostgresSourceSpecificMetrics {
     }
 }
 
+/// Metrics for the `upsert` operator.
+#[derive(Clone, Debug)]
+pub(super) struct UpsertMetrics {
+    pub(super) rehydration_latency: GaugeVec,
+    pub(super) rehydration_total: UIntGaugeVec,
+
+    // These are used by `shared`.
+    pub(super) multi_get_latency: HistogramVec,
+    pub(super) multi_get_size: HistogramVec,
+    pub(super) multi_put_latency: HistogramVec,
+    pub(super) multi_put_size: HistogramVec,
+    // This is a map so that multiple timely workers can interact with the same
+    // `DeleteOnDropHistogram`, which is only dropped once ALL workers drop it.
+    // The map may contain arbitrary, old `Weak`s for deleted sources, which are
+    // only cleaned if those sources are recreated.
+    //
+    // We don't parameterize these by `worker_id` like the `rehydration_*` ones
+    // to save on time-series cardinality.
+    pub(super) shared: Arc<Mutex<BTreeMap<GlobalId, Weak<UpsertSharedMetrics>>>>,
```

Comment:
Can each worker have its private clone of `Metrics` that point to the same labels? What does the `Arc<Mutex<` buy us? I'm probably misisng a lot of stuff for how metrics are wired up so I'll ask some follow up questions in slack

Reply:
Its very complicated, and i was forced to do it this way! Feel free to ask in slack!